### ISSUE TITLE
Update provider, use map as input rather than object

### ISSUE THREAD

**Issue by Sriep** [0chain/0chain#2544]
Update provider smart contract endpoints currently use an object as input. This can cause an issue distinguishing between unchanged and default settings. 

A suggestion; we change the input to the update_provider endpoints to take a map of updated fields as input rather than an object. 

Will need `zbox` ad `zwallet` changes as well.

**Comment by Sriep** [comment]
We want to modify the endpoints that update providers. For example [update_blobber_settings](https://github.com/0chain/0chain/blob/staging/code/go/0chain.net/smartcontract/storagesc/sc.go#L214) and [update_validator_settings](https://github.com/0chain/0chain/blob/staging/code/go/0chain.net/smartcontract/storagesc/sc.go#L216), there should be update miner, updare sharder and update authoriser endpoints as well.

https://github.com/0chain/0chain/blob/staging/code/go/0chain.net/smartcontract/storagesc/blobber.go#L212
Here the input is seen to be a blobber object. The problem is that if a field is the golang default, we can't tell if we are setting it or ignoring it. There was a problem with setting the service charge to zero by accident. We want to change to a map[stirng]interface{}.

We will also want to changes to goSDK, zbox and zwallet code.  For example
https://github.com/0chain/zwalletcli/blob/staging/cmd/updatenode.go#L105
```go
		if sharder {
			err = txn.MinerSCSharderSettings(miner)
		} else {
			err = txn.MinerSCMinerSettings(miner)
		}
```
here the MinerSC*Settings method will take a map of changes rather than an object. The function is in goSDK so that will need changing. We should add missing fields here as well - currently, you cannot set the service charge so always set it to zero - clearly a bug. 

Miners, Sharders, Blobbers, Validators and authorised should all work the same way. Modify the existing code to use maps rather than objects.

**Comment by sunilmhta** [comment]
Hi @dabasov, @Sriep I went through the code and tried to make sense of our use cases. There are two ways to do this as of now:
### 1. Using Maps:
If we use maps, we would need to define something like this 

#### From caller side:
``` go
changedSettings := make(map[string]interface{})
changedSettings["min_stake"] = newMinStakeInAppropriateDataType
txn.MinerSCSharderSettings(changedSettings)
```
#### From the consumer side:
``` go
func (t *Transaction) MinerSCSharderSettings(info *MinerSCMinerInfo) (err error) {}
```

-  We would need to change the argument type of this method from `MinerSCMinerInfo` to something else.
-	After deserializing to json, we would also need to map the keys of the map to the corresponding value and populate them. I do not see any clear way apart from some messy `switch` `case` statements and a `for` loop. There might be some other ways using `reflect` but that will probably add overhead and I do not think that's okay.

### 2. Using `omitempty` with pointer types:

We could probably refactor the existing structs and convert the basic types to pointer types and check for `nil` something like this:
``` go
type StakePoolSettings struct {
	DelegateWallet *string         `json:"delegate_wallet,omitempty"`
	MinStake       *common.Balance `json:"min_stake,omitempty"`
	MaxStake       *common.Balance `json:"max_stake,omitempty"`
	NumDelegates   *int            `json:"num_delegates,omitempty"`
	ServiceCharge  *float64        `json:"service_charge,omitempty"`
}
```
Now we can safely distinguish the two cases where a field is initialized with their zero value or it is non initialized like this:
``` go
serviceCharge := 0.0
stakePoolSettings := StakePoolSettings {
	ServiceCharge : &serviceCharge,
}
```
In the above object, we will know if `stakePoolSettings.ServiceCharge` is actually `0.00` or is it the zero value.

I prefer the second approach. What do you guys think?

**Comment by dabasov** [comment]
I like omitempty approach more than maps, the main advantage of the former is that we keep tag name checking on client and server, it's not real validation, but it's better than nothing in case of maps, where we can misspell the tag name, also types are validated automatically

**Comment by Kishan-Dhakan** [comment]
Closed by #2577

**Comment by sunilmhta** [comment]
This issue is not closed yet, there are one more set of PRs coming. I will close it manually. Please do not close it before that.
